### PR TITLE
Backend refactor

### LINKED ISSUE #103: Board.discover is redundant
I can't think of any situations where `board.discover(backend)` would do anything except simply call `backend.discover()`, so the former could likely be removed in favour of calling the latter directly.

### PR DESCRIPTION
**Pull request by trickeydan**
Fixes #103, #116

### CODE REVIEW

**Comment by trickeydan** [comment]
I've concluded that a `Backend` shouldn't actually need any references to the Board it is controlling. All required data to perform it's actions should be passed in the constructor.

File: tests/backends/hardware/sr/v4/test_power_board.py [hunk 1]
```diff
@@ -354,10 +354,10 @@ def test_backend_get_power_output_enabled():
     backend = SRV4PowerBoardHardwareBackend(device)
 
     for i in range(0, 6):
-        assert not backend.get_power_output_enabled(MockBoard(), i)
```
Comment (kierdavis) [inline tests/backends/hardware/sr/v4/test_power_board.py]: Please delete the `MockBoard` class, if it is now unused.

**Comment by trickeydan** [comment]
There are a lot of changes in this PR. I'd recommend review commit by commit.

File: j5/backends/hardware/sr/v4/power_board.py [hunk 2]
```diff
@@ -234,56 +233,52 @@ def get_power_output_current(self, board: Board, identifier: int) -> float:
         current, = struct.unpack("<I", self._read(cmd))
         return cast(int, current) / 1000  # convert milliamps to amps
 
-    def buzz(self, board: Board, identifier: int,
-             duration: timedelta, pitch: Pitch) -> None:
+    def buzz(self, identifier: int,
+             duration: timedelta, pitch: int) -> None:
```
Comment (kierdavis) [inline j5/backends/hardware/sr/v4/power_board.py]: If this function now only takes a frequency and no longer accepts a `Note`, please can rename this argument from `pitch` to `frequency`?

File: j5/boards/board.py [hunk 3]
```diff
@@ -78,55 +72,70 @@ def make_all_safe() -> None:
...
+        """The type of board that this group contains."""
+        return self._board_class
+
+    @property
+    def backend_class(self) -> Type[Backend]:
+        """The Backend that this group uses for Boards."""
+        return self._backend_class
+
+    @property
+    def boards(self) -> List[Board]:
```
Comment (kierdavis) [inline j5/boards/board.py]: I think a `Set` would be a more appropriate return type for this function.

Comment (trickeydan) [reply j5/boards/board.py]: Split out into #207

File: j5/backends/hardware/sr/v4/power_board.py (hunk 2 above)
Comment (trickeydan) [reply j5/backends/hardware/sr/v4/power_board.py]: This change was made in a separate PR, but I'll fix it here anyway

**Pull request merged**